### PR TITLE
Pro 1522 integration setup tesla

### PR DESCRIPTION
**Pull request by Allyson-English**

### CODE REVIEW

**Comment by elffjs** [comment]
Do we know why the tests are failing?

File: internal/controllers/synthetic_devices_controller.go [hunk 1]
```diff
@@ -433,6 +431,45 @@ func (vc *SyntheticDevicesController) handleDeviceAPIIntegrationCreation(ctx con
...
+		}
+		encRefresh, err := vc.cipher.Encrypt(req.Credentials.RefreshToken)
+		if err != nil {
+			return opaqueInternalError
+		}
+		udi.AccessToken = null.StringFrom(encAccess)
+		udi.AccessExpiresAt = null.TimeFrom(time.Unix(req.Credentials.ExpiresIn, 0))
+		udi.RefreshToken = null.StringFrom(encRefresh)
+
+		meta := services.UserDeviceAPIIntegrationsMetadata{
```
Comment (elffjs) [inline internal/controllers/synthetic_devices_controller.go:458]: Good catch.

File: internal/controllers/synthetic_devices_controller.go [hunk 2]
```diff
@@ -433,6 +431,45 @@ func (vc *SyntheticDevicesController) handleDeviceAPIIntegrationCreation(ctx con
...
+		encAccess, err := vc.cipher.Encrypt(req.Credentials.AccessToken)
+		if err != nil {
+			return opaqueInternalError
+		}
+		encRefresh, err := vc.cipher.Encrypt(req.Credentials.RefreshToken)
+		if err != nil {
+			return opaqueInternalError
+		}
+		udi.AccessToken = null.StringFrom(encAccess)
+		udi.AccessExpiresAt = null.TimeFrom(time.Unix(req.Credentials.ExpiresIn, 0))
```
Comment (elffjs) [inline internal/controllers/synthetic_devices_controller.go:455]: In the OAuth2 spec, `expires_in` is the amount of time you have until the credentials expire, in seconds starting from "now". So we probably want to do some addition here.

File: internal/controllers/synthetic_devices_controller.go [hunk 3]
```diff
@@ -264,33 +271,24 @@ func (vc *SyntheticDevicesController) MintSyntheticDevice(c *fiber.Ctx) error {
 
 	rawPayload := vc.getEIP712Mint(int64(integration.TokenId), vid)
 
-	hash, _, err := signer.TypedDataAndHash(*rawPayload)
+	h, _, err := signer.TypedDataAndHash(*rawPayload)
 	if err != nil {
 		vc.log.Err(err).Msg("Error occurred creating has of payload")
 		return fiber.NewError(fiber.StatusBadRequest, "Couldn't verify signature.")
 	}
+	hash := crypto.Keccak256Hash(h)
```
Comment (elffjs) [inline internal/controllers/synthetic_devices_controller.go:279]: I'm worried about `h` and `hash` here. I think `TypedDataAndHash` [is already hashing](TypedDataAndHash). They don't return a `common.Hash` for whatever reason. I'd be shocked if line 282 succeeded.

File: internal/controllers/synthetic_devices_controller.go [hunk 4]
```diff
@@ -264,33 +271,24 @@ func (vc *SyntheticDevicesController) MintSyntheticDevice(c *fiber.Ctx) error {
...
+	recAddr, err := helpers.Ecrecover(hash, ownerSignature)
 	if err != nil {
-		vc.log.Err(err).Msg("Error occurred marshalling recovered public public key")
-		return fiber.NewError(fiber.StatusBadRequest, "Couldn't verify signature.")
+		vc.log.Err(err).Msg("unable to validate signature")
+		return err
 	}
-
-	payloadVerified := bytes.Equal(crypto.PubkeyToAddress(*pubRaw).Bytes(), userAddr.Bytes())
+	payloadVerified := bytes.Equal(recAddr.Bytes(), userAddr.Bytes())
```
Comment (elffjs) [inline internal/controllers/synthetic_devices_controller.go:287]: You can just check if they're equal!

File: internal/controllers/synthetic_devices_controller.go [hunk 5]
```diff
@@ -264,33 +271,24 @@ func (vc *SyntheticDevicesController) MintSyntheticDevice(c *fiber.Ctx) error {
...
 	if !payloadVerified {
 		return fiber.NewError(fiber.StatusBadRequest, "Invalid signature provided")
 	}
 
-	ownerSignature[64] += 27
```
Comment (elffjs) [inline internal/controllers/synthetic_devices_controller.go:292]: Nice to be able to dump this.

File: internal/controllers/synthetic_devices_controller.go (hunk 4 above)
Comment (Allyson-English) [reply internal/controllers/synthetic_devices_controller.go:287]: gave `recAddr.Bytes() == userAddr.Bytes()` a try-- looks like recAddr.Bytes() is returning a slice.. did you having something else in mind?

Comment (elffjs) [reply internal/controllers/synthetic_devices_controller.go:286]: `recAddr` is an address, `userAddr` is an address.

File: internal/controllers/synthetic_devices_controller.go [hunk 6]
```diff
@@ -261,36 +268,26 @@ func (vc *SyntheticDevicesController) MintSyntheticDevice(c *fiber.Ctx) error {
 	}
 
 	userAddr := common.HexToAddress(*user.EthereumAddress)
-
 	rawPayload := vc.getEIP712Mint(int64(integration.TokenId), vid)
 
-	hash, _, err := signer.TypedDataAndHash(*rawPayload)
+	h, _, err := signer.TypedDataAndHash(*rawPayload)
```
Comment (elffjs) [inline internal/controllers/synthetic_devices_controller.go:273]: We can probably bring back a good name.

File: internal/controllers/synthetic_devices_controller.go [hunk 7]
```diff
@@ -433,6 +430,45 @@ func (vc *SyntheticDevicesController) handleDeviceAPIIntegrationCreation(ctx con
 		mb, _ := json.Marshal(meta)
 		udi.Metadata = null.JSONFrom(mb)
 		udi.ExternalID = null.StringFrom(externalID)
+	case constants.TeslaVendor:
+		teslaID, err := strconv.Atoi(req.Credentials.ExternalID)
+		if err != nil {
+			vc.log.Err(err).Msgf("unable to parse external id %+v as integer", req.Credentials.ExternalID)
```
Comment (elffjs) [inline internal/controllers/synthetic_devices_controller.go:436]: I don't see why we'd need to do `%+v`.

File: internal/controllers/synthetic_devices_controller.go (hunk 2 above)
Comment (elffjs) [reply internal/controllers/synthetic_devices_controller.go:454]: Why are we setting this twice? That's odd.

File: internal/controllers/synthetic_devices_controller.go [hunk 8]
```diff
@@ -261,36 +268,26 @@ func (vc *SyntheticDevicesController) MintSyntheticDevice(c *fiber.Ctx) error {
...
 
 	userAddr := common.HexToAddress(*user.EthereumAddress)
-
 	rawPayload := vc.getEIP712Mint(int64(integration.TokenId), vid)
 
-	hash, _, err := signer.TypedDataAndHash(*rawPayload)
+	h, _, err := signer.TypedDataAndHash(*rawPayload)
 	if err != nil {
-		vc.log.Err(err).Msg("Error occurred creating has of payload")
+		vc.log.Err(err).Msg("Error occurred creating hash of payload")
```
Comment (elffjs) [inline internal/controllers/synthetic_devices_controller.go:275]: 👍

**Pull request merged**